state/entrysync.go

```diff
@@ -0,0 +1,313 @@
+//
+// Check:
+// 1. Look at the first record of the eblock queue
+// 2. Go through all missing records and check if they exist
+// 3. If there are still missing records, wait and go back to 1
+// 4. If the eblock is finished, inform the state we have all the entries up to that eblock
+type EntrySync struct {
+	s *State
+
+	position uint32
+
+	askMtx sync.RWMutex
+	askMap map[[32]byte]bool // entry hash => ask
+	asks   []*entrySyncAsk   // chronological
+
+	ebMtx   sync.Mutex
+	eblocks []*entrySyncEBlock
+
+	closer chan interface{}
+}
+
+// an item in the ask queue
+type entrySyncAsk struct {
+	hash  interfaces.IHash
+	time  time.Time
+	tries int
+}
+
+// an item in the eblock queue
+type entrySyncEBlock struct {
+	height  uint32
+	missing []interfaces.IHash
+}
+
+// NewEntrySync creates a new EntrySync
+func NewEntrySync(s *State) *EntrySync {
+	es := new(EntrySync)
+	es.s = s
+	es.closer = make(chan interface{}, 1)
+	es.askMap = make(map[[32]byte]bool)
+	return es
+}
+
+// Stop initiates shutdown and will stop all associated goroutines
+func (es *EntrySync) Stop() {
+	close(es.closer)
+}
+
+// check routine, for description see EntrySync comment
+func (es *EntrySync) check() {
+	for {
+		select {
+		case <-es.closer:
+			return
+		default:
+		}
+
+		if len(es.eblocks) == 0 { // outside of mutex but should be fine
+			time.Sleep(time.Second)
+			continue
+		}
+
+		es.ebMtx.Lock()
+		eb := es.eblocks[0]
+		es.s.SetEntryBlockDBHeightProcessing(eb.height)
+
+		stillmissing := make([]interfaces.IHash, 0, len(eb.missing))
+		for _, entryhash := range eb.missing {
+			if !es.has(entryhash) {
+				stillmissing = append(stillmissing, entryhash)
+			}
+		}
+
+		if len(stillmissing) > 0 {
+			eb.missing = stillmissing
+		} else { // eblock is complete
+			es.eblocks = es.eblocks[1:]
+			es.s.SetEntryBlockDBHeightComplete(eb.height)
+		}
+		es.ebMtx.Unlock()
+
+		if len(stillmissing) > 0 { // waiting on requests
+			time.Sleep(time.Second)
+		}
+	}
+}
+
+// has checks if the hash exists in the database
+func (es *EntrySync) has(hash interfaces.IHash) bool {
+	if has, err := es.s.DB.DoesKeyExist(databaseOverlay.ENTRY, hash.Bytes()); err != nil {
+		panic(err)
+	} else {
+		return has
+	}
+}
+
+// ask routine, for description see EntrySync comment
+func (es *EntrySync) ask() {
+	for {
+		select {
+		case <-es.closer:
+			return
+		default:
+		}
+
+		es.askMtx.Lock()
+		// check the first N entries in the list
+		for i := 0; i < len(es.asks) && i < EntrySyncMax; i++ {
+			ask := es.asks[i]
+
+			// remove asks that are fulfilled
+			if es.has(ask.hash) {
+				es.asks = append(es.asks[:i], es.asks[i+1:]...)
+				i-- // we removed one, index stays the same
+				delete(es.askMap, ask.hash.Fixed())
+				continue
+			}
+
+			// re-ask
+			if time.Since(ask.time) > EntrySyncRetry {
+				request := messages.NewMissingData(primitives.NewTimestampNow(), ask.hash)
+				request.SendOut(es.s, request)
```

Comment:
Just to confirm, this is going to be sent to a single random peer right?

Reply:
Yup. NewMissingData sets the flag "Peer2Peer" to true:
https://github.com/FactomProject/factomd/blob/ba6800ab3ba2296acf82726af2bd64d65c88b7ca/common/messages/missingData.go#L219

and the P2PProxy takes a "P2P Message with no PeerHash" and sends it to a random peer:
https://github.com/FactomProject/factomd/blob/ba6800ab3ba2296acf82726af2bd64d65c88b7ca/engine/p2pProxy.go#L135-L137